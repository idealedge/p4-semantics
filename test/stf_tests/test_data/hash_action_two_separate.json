{
  "program" : "hash_action_two_separate.p4",
  "__meta__" : {
    "version" : [2, 7],
    "compiler" : "https://github.com/p4lang/p4c"
  },
  "header_types" : [
    {
      "name" : "scalars_0",
      "id" : 0,
      "fields" : [
        ["tmp", 32, false],
        ["tmp_0", 32, false]
      ]
    },
    {
      "name" : "data_t",
      "id" : 1,
      "fields" : [
        ["f1", 32, false],
        ["f2", 32, false],
        ["h1", 16, false],
        ["h2", 16, false],
        ["h3", 16, false],
        ["h4", 16, false]
      ]
    },
    {
      "name" : "counter_metadata_t",
      "id" : 2,
      "fields" : [
        ["counter_index_first", 16, false],
        ["counter_index_second", 16, false]
      ]
    },
    {
      "name" : "standard_metadata",
      "id" : 3,
      "fields" : [
        ["ingress_port", 9, false],
        ["egress_spec", 9, false],
        ["egress_port", 9, false],
        ["clone_spec", 32, false],
        ["instance_type", 32, false],
        ["drop", 1, false],
        ["recirculate_port", 16, false],
        ["packet_length", 32, false],
        ["enq_timestamp", 32, false],
        ["enq_qdepth", 19, false],
        ["deq_timedelta", 32, false],
        ["deq_qdepth", 19, false],
        ["ingress_global_timestamp", 48, false],
        ["lf_field_list", 32, false],
        ["mcast_grp", 16, false],
        ["resubmit_flag", 1, false],
        ["egress_rid", 16, false],
        ["checksum_error", 1, false],
        ["_padding", 4, false]
      ]
    }
  ],
  "headers" : [
    {
      "name" : "scalars",
      "id" : 0,
      "header_type" : "scalars_0",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "standard_metadata",
      "id" : 1,
      "header_type" : "standard_metadata",
      "metadata" : true,
      "pi_omit" : true
    },
    {
      "name" : "data",
      "id" : 2,
      "header_type" : "data_t",
      "metadata" : false,
      "pi_omit" : true
    },
    {
      "name" : "counter_metadata",
      "id" : 3,
      "header_type" : "counter_metadata_t",
      "metadata" : true,
      "pi_omit" : true
    }
  ],
  "header_stacks" : [],
  "header_union_types" : [],
  "header_unions" : [],
  "header_union_stacks" : [],
  "field_lists" : [],
  "errors" : [
    ["NoError", 1],
    ["PacketTooShort", 2],
    ["NoMatch", 3],
    ["StackOutOfBounds", 4],
    ["HeaderTooShort", 5],
    ["ParserTimeout", 6]
  ],
  "enums" : [],
  "parsers" : [
    {
      "name" : "parser",
      "id" : 0,
      "init_state" : "start",
      "parse_states" : [
        {
          "name" : "start",
          "id" : 0,
          "parser_ops" : [
            {
              "parameters" : [
                {
                  "type" : "regular",
                  "value" : "data"
                }
              ],
              "op" : "extract"
            }
          ],
          "transitions" : [
            {
              "value" : "default",
              "mask" : null,
              "next_state" : null
            }
          ],
          "transition_key" : []
        }
      ]
    }
  ],
  "deparsers" : [
    {
      "name" : "deparser",
      "id" : 0,
      "order" : ["data"]
    }
  ],
  "meter_arrays" : [],
  "counter_arrays" : [
    {
      "name" : "count1",
      "id" : 0,
      "size" : 16384,
      "is_direct" : false
    },
    {
      "name" : "count2",
      "id" : 1,
      "size" : 16384,
      "is_direct" : false
    }
  ],
  "register_arrays" : [],
  "calculations" : [],
  "learn_lists" : [],
  "actions" : [
    {
      "name" : "NoAction",
      "id" : 0,
      "runtime_data" : [],
      "primitives" : []
    },
    {
      "name" : "set_index",
      "id" : 1,
      "runtime_data" : [
        {
          "name" : "index1",
          "bitwidth" : 16
        },
        {
          "name" : "index2",
          "bitwidth" : 16
        },
        {
          "name" : "port",
          "bitwidth" : 9
        }
      ],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["counter_metadata", "counter_index_first"]
            },
            {
              "type" : "runtime_data",
              "value" : 0
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 29,
            "column" : 17,
            "source_fragment" : "index1, index2, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["counter_metadata", "counter_index_second"]
            },
            {
              "type" : "runtime_data",
              "value" : 1
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 29,
            "column" : 25,
            "source_fragment" : "index2, port) { ..."
          }
        },
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["standard_metadata", "egress_spec"]
            },
            {
              "type" : "runtime_data",
              "value" : 2
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 29,
            "column" : 33,
            "source_fragment" : "port) { ..."
          }
        }
      ]
    },
    {
      "name" : "count_entries",
      "id" : 2,
      "runtime_data" : [],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["scalars", "tmp"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "&",
                  "left" : {
                    "type" : "field",
                    "value" : ["counter_metadata", "counter_index_first"]
                  },
                  "right" : {
                    "type" : "hexstr",
                    "value" : "0xffffffff"
                  }
                }
              }
            }
          ]
        },
        {
          "op" : "count",
          "parameters" : [
            {
              "type" : "counter_array",
              "value" : "count1"
            },
            {
              "type" : "field",
              "value" : ["scalars", "tmp"]
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 61,
            "column" : 4,
            "source_fragment" : "count(count1, counter_metadata.counter_index_first)"
          }
        }
      ]
    },
    {
      "name" : "count_entries2",
      "id" : 3,
      "runtime_data" : [],
      "primitives" : [
        {
          "op" : "assign",
          "parameters" : [
            {
              "type" : "field",
              "value" : ["scalars", "tmp_0"]
            },
            {
              "type" : "expression",
              "value" : {
                "type" : "expression",
                "value" : {
                  "op" : "&",
                  "left" : {
                    "type" : "field",
                    "value" : ["counter_metadata", "counter_index_second"]
                  },
                  "right" : {
                    "type" : "hexstr",
                    "value" : "0xffffffff"
                  }
                }
              }
            }
          ]
        },
        {
          "op" : "count",
          "parameters" : [
            {
              "type" : "counter_array",
              "value" : "count2"
            },
            {
              "type" : "field",
              "value" : ["scalars", "tmp_0"]
            }
          ],
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 72,
            "column" : 4,
            "source_fragment" : "count(count2, counter_metadata.counter_index_second)"
          }
        }
      ]
    }
  ],
  "pipelines" : [
    {
      "name" : "ingress",
      "id" : 0,
      "init_table" : "index_setter",
      "tables" : [
        {
          "name" : "index_setter",
          "id" : 0,
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 35,
            "column" : 0,
            "source_fragment" : "table index_setter { ..."
          },
          "key" : [
            {
              "match_type" : "exact",
              "target" : ["data", "f1"],
              "mask" : null
            },
            {
              "match_type" : "exact",
              "target" : ["data", "f2"],
              "mask" : null
            }
          ],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 2048,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [1, 0],
          "actions" : ["set_index", "NoAction"],
          "base_default_next" : "stats",
          "next_tables" : {
            "set_index" : "stats",
            "NoAction" : "stats"
          },
          "default_entry" : {
            "action_id" : 0,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "stats",
          "id" : 1,
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 64,
            "column" : 0,
            "source_fragment" : "table stats { ..."
          },
          "key" : [],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [2],
          "actions" : ["count_entries"],
          "base_default_next" : "stats2",
          "next_tables" : {
            "count_entries" : "stats2"
          },
          "default_entry" : {
            "action_id" : 2,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        },
        {
          "name" : "stats2",
          "id" : 2,
          "source_info" : {
            "filename" : "hash_action_two_separate.p4",
            "line" : 75,
            "column" : 0,
            "source_fragment" : "table stats2 { ..."
          },
          "key" : [],
          "match_type" : "exact",
          "type" : "simple",
          "max_size" : 1024,
          "with_counters" : false,
          "support_timeout" : false,
          "direct_meters" : null,
          "action_ids" : [3],
          "actions" : ["count_entries2"],
          "base_default_next" : null,
          "next_tables" : {
            "count_entries2" : null
          },
          "default_entry" : {
            "action_id" : 3,
            "action_const" : false,
            "action_data" : [],
            "action_entry_const" : false
          }
        }
      ],
      "action_profiles" : [],
      "conditionals" : []
    },
    {
      "name" : "egress",
      "id" : 1,
      "init_table" : null,
      "tables" : [],
      "action_profiles" : [],
      "conditionals" : []
    }
  ],
  "checksums" : [],
  "force_arith" : [],
  "extern_instances" : [],
  "field_aliases" : [
    [
      "queueing_metadata.enq_timestamp",
      ["standard_metadata", "enq_timestamp"]
    ],
    [
      "queueing_metadata.enq_qdepth",
      ["standard_metadata", "enq_qdepth"]
    ],
    [
      "queueing_metadata.deq_timedelta",
      ["standard_metadata", "deq_timedelta"]
    ],
    [
      "queueing_metadata.deq_qdepth",
      ["standard_metadata", "deq_qdepth"]
    ],
    [
      "intrinsic_metadata.ingress_global_timestamp",
      ["standard_metadata", "ingress_global_timestamp"]
    ],
    [
      "intrinsic_metadata.lf_field_list",
      ["standard_metadata", "lf_field_list"]
    ],
    [
      "intrinsic_metadata.mcast_grp",
      ["standard_metadata", "mcast_grp"]
    ],
    [
      "intrinsic_metadata.resubmit_flag",
      ["standard_metadata", "resubmit_flag"]
    ],
    [
      "intrinsic_metadata.egress_rid",
      ["standard_metadata", "egress_rid"]
    ]
  ]
}