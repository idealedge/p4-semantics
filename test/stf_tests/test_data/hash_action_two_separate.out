<T> <k> @nextPacket </k> <headers> <header> <name> %standard_metadata_t </name> <opts> .Map </opts> <fields> egress_spec : 32 ( .FieldMods ) ;  ( egress_spec : 32 ( .FieldMods ) ;  .FieldDecs ) </fields> </header> <header> <name> counter_metadata_t </name> <opts> "$fixed_width" |-> 32 </opts> <fields> counter_index_first : 16 ( .FieldMods ) ;  ( counter_index_second : 16 ( .FieldMods ) ;  .FieldDecs ) </fields> </header> <header> <name> data_t </name> <opts> "$fixed_width" |-> 128 </opts> <fields> f1 : 32 ( .FieldMods ) ;  ( f2 : 32 ( .FieldMods ) ;  ( h1 : 16 ( .FieldMods ) ;  ( h2 : 16 ( .FieldMods ) ;  ( h3 : 16 ( .FieldMods ) ;  ( h4 : 16 ( .FieldMods ) ;  .FieldDecs ) ) ) ) ) </fields> </header> </headers> <arrays> .Map </arrays> <fieldLists> .FieldListCellBag </fieldLists> <fieldListCalcs> .FieldListCalcCellBag </fieldListCalcs> <calcFields> .CalcFieldCellBag </calcFields> <instances> <instance> <valid> true </valid> <metadata> true </metadata> <typeName> %standard_metadata_t </typeName> <name> standard_metadata </name> <fieldVals> ingress_port |-> @val ( 0 , 32 , false ) egress_spec |-> @val ( 3 , 32 , false ) </fieldVals> </instance> <instance> <valid> true </valid> <metadata> true </metadata> <typeName> counter_metadata_t </typeName> <name> counter_metadata </name> <fieldVals> counter_index_first |-> @val ( 8 , 16 , false ) counter_index_second |-> @val ( 11 , 16 , false ) </fieldVals> </instance> <instance> <valid> true </valid> <metadata> false </metadata> <typeName> data_t </typeName> <name> data </name> <fieldVals> h2 |-> @val ( 1285 , 16 , false ) h1 |-> @val ( 1028 , 16 , false ) h4 |-> @val ( 1799 , 16 , false ) f2 |-> @val ( 1431655765 , 32 , false ) h3 |-> @val ( 1542 , 16 , false ) f1 |-> @val ( 1145324612 , 32 , false ) </fieldVals> </instance> </instances> <initInstances> <instances>-fragment <instance> <valid> true </valid> <metadata> true </metadata> <typeName> %standard_metadata_t </typeName> <name> standard_metadata </name> <fieldVals> ingress_port |-> @undef egress_spec |-> @undef </fieldVals> </instance> <instance> <valid> true </valid> <metadata> true </metadata> <typeName> counter_metadata_t </typeName> <name> counter_metadata </name> <fieldVals> counter_index_first |-> @val ( 0 , 16 , false ) counter_index_second |-> @val ( 0 , 16 , false ) </fieldVals> </instance> <instance> <valid> false </valid> <metadata> false </metadata> <typeName> data_t </typeName> <name> data </name> <fieldVals> .Map </fieldVals> </instance> </instances>-fragment </initInstances> <initArrays> .Map </initArrays> <parserStates> <state> <name> start </name> <body> ( extract ( data ) ; .ExtractOrSetStatements ) return ingress ; </body> </state> </parserStates> <valueSets> .ValueSetCellBag </valueSets> <exceptionHandlers> .ExceptionHandlerCellBag </exceptionHandlers> <handledExceptions> .Set </handledExceptions> <statefuls> <stateful> <stype> $counter </stype> <name> count2 </name> <opts> "$instant_count" |-> 16384 "$table" |-> stats2 "$signed" |-> false "$binding" |-> $static "$counter_type" |-> packets "$saturating" |-> false "$min_width" |-> 32 </opts> <vals> .Map </vals> <width> 0 </width> </stateful> <stateful> <stype> $counter </stype> <name> count1 </name> <opts> "$instant_count" |-> 16384 "$table" |-> stats "$signed" |-> false "$binding" |-> $static "$counter_type" |-> packets "$saturating" |-> false "$min_width" |-> 32 </opts> <vals> .Map </vals> <width> 0 </width> </stateful> </statefuls> <tableDirects> .Map </tableDirects> <actions> <action> <name> count_entries2 </name> <params> .ParamList </params> <body> count ( count2 , ( ( counter_metadata . counter_index_second ) , .Args ) ) ;  .ActionStatements </body> </action> <action> <name> count_entries </name> <params> .ParamList </params> <body> count ( count1 , ( ( counter_metadata . counter_index_first ) , .Args ) ) ;  .ActionStatements </body> </action> <action> <name> set_index </name> <params> index1 , ( index2 , ( port , .ParamList ) ) </params> <body> modify_field ( ( counter_metadata . counter_index_first ) , ( index1 , .Args ) ) ;  ( modify_field ( ( counter_metadata . counter_index_second ) , ( index2 , .Args ) ) ;  ( modify_field ( ( standard_metadata . egress_spec ) , ( port , .Args ) ) ;  .ActionStatements ) ) </body> </action> </actions> <actionProfiles> .ActionProfileCellBag </actionProfiles> <actionSelectors> .ActionSelectorCellBag </actionSelectors> <tables> <table> <name> stats2 </name> <reads> .FieldMatchs </reads> <acts> actions { count_entries2 ;  .ActionNameItems } </acts> <opts> .TableOptionals </opts> <rules> .List </rules> <default> . </default> </table> <table> <name> stats </name> <reads> .FieldMatchs </reads> <acts> actions { count_entries ;  .ActionNameItems } </acts> <opts> .TableOptionals </opts> <rules> .List </rules> <default> . </default> </table> <table> <name> index_setter </name> <reads> ( data . f1 ) : exact ;  ( ( data . f2 ) : exact ;  .FieldMatchs ) </reads> <acts> actions { set_index ;  .ActionNameItems } </acts> <opts> size : 2048 ;  .TableOptionals </opts> <rules> ListItem ( $rule ( 0 , $ctr ( ListItem ( @val ( 286331153 , 0 , false ) ) ListItem ( @val ( 572662306 , 0 , false ) ) ) , @call ( set_index , $resolved ( ListItem ( @val ( 9 , 0 , false ) ) ListItem ( @val ( 10 , 0 , false ) ) ListItem ( @val ( 1 , 0 , false ) ) ) ) ) ) ListItem ( $rule ( 1 , $ctr ( ListItem ( @val ( 572662306 , 0 , false ) ) ListItem ( @val ( 858993459 , 0 , false ) ) ) , @call ( set_index , $resolved ( ListItem ( @val ( 9 , 0 , false ) ) ListItem ( @val ( 11 , 0 , false ) ) ListItem ( @val ( 2 , 0 , false ) ) ) ) ) ) ListItem ( $rule ( 2 , $ctr ( ListItem ( @val ( 858993459 , 0 , false ) ) ListItem ( @val ( 1145324612 , 0 , false ) ) ) , @call ( set_index , $resolved ( ListItem ( @val ( 9 , 0 , false ) ) ListItem ( @val ( 11 , 0 , false ) ) ListItem ( @val ( 2 , 0 , false ) ) ) ) ) ) ListItem ( $rule ( 3 , $ctr ( ListItem ( @val ( 1145324612 , 0 , false ) ) ListItem ( @val ( 1431655765 , 0 , false ) ) ) , @call ( set_index , $resolved ( ListItem ( @val ( 8 , 0 , false ) ) ListItem ( @val ( 11 , 0 , false ) ) ListItem ( @val ( 3 , 0 , false ) ) ) ) ) ) </rules> <default> . </default> </table> </tables> <controlFlows> <control> <name> ingress </name> <body> apply ( index_setter ) { .HitMissCases }  ( apply ( stats ) { .HitMissCases }  ( apply ( stats2 ) { .HitMissCases }  .ControlStatements ) ) </body> </control> </controlFlows> <cfset> SetItem ( ingress ) </cfset> <calcFieldSet> .Set </calcFieldSet> <ctx> <frameStack> .List </frameStack> <crnt> "$rule" |-> . "$table" |-> . "$action" |-> set_index "$miss" |-> true </crnt> </ctx> <packet> "01000100010001000100010001000100010101010101010101010101010101010000010000000100000001010000010100000110000001100000011100000111" </packet> <packetout> .List </packetout> <parser> <graph> <onpath> .Set </onpath> <marked> <mstates> SetItem ( start ) </mstates> <mheaders> SetItem ( data ) </mheaders> </marked> <dporder> ListItem ( data ) </dporder> </graph> <pctx> <index> 128 </index> <lastExt> data </lastExt> <varWidth> . </varWidth> <packetSize> @val ( 16 , 0 , false ) </packetSize> </pctx> </parser> <buffer> <in> .List </in> <out> ListItem ( $packet ( ListItem ( @val ( 286331153 , 32 , false ) ) ListItem ( @val ( 572662306 , 32 , false ) ) ListItem ( @val ( 257 , 16 , false ) ) ListItem ( @val ( 514 , 16 , false ) ) ListItem ( @val ( 771 , 16 , false ) ) ListItem ( @val ( 1028 , 16 , false ) ) , 1 ) ) ListItem ( $packet ( ListItem ( @val ( 572662306 , 32 , false ) ) ListItem ( @val ( 858993459 , 32 , false ) ) ListItem ( @val ( 514 , 16 , false ) ) ListItem ( @val ( 771 , 16 , false ) ) ListItem ( @val ( 1028 , 16 , false ) ) ListItem ( @val ( 1285 , 16 , false ) ) , 2 ) ) ListItem ( $packet ( ListItem ( @val ( 858993459 , 32 , false ) ) ListItem ( @val ( 1145324612 , 32 , false ) ) ListItem ( @val ( 771 , 16 , false ) ) ListItem ( @val ( 1028 , 16 , false ) ) ListItem ( @val ( 1285 , 16 , false ) ) ListItem ( @val ( 1542 , 16 , false ) ) , 2 ) ) ListItem ( $packet ( ListItem ( @val ( 1145324612 , 32 , false ) ) ListItem ( @val ( 1431655765 , 32 , false ) ) ListItem ( @val ( 1028 , 16 , false ) ) ListItem ( @val ( 1285 , 16 , false ) ) ListItem ( @val ( 1542 , 16 , false ) ) ListItem ( @val ( 1799 , 16 , false ) ) , 3 ) ) </out> </buffer> </T>
